lib/setMode.js

```
@@ -1,21 +1,160 @@
+
+    //Send color data for right side
+    sendData(keyboard, 7, rightDef.primary.color, rightDef.primary.level, 0);
+    sendData(keyboard, 8, rightDef.secondary.color, rightDef.secondary.level, 0);
+    sendData(keyboard, 9, period, period, period);
+  }
+  
+  //Initalize the mode
+  commit(keyboard, constants.modes[mode]);
+  
+  //Save the mode for reference
+  keyboard.currentMode = mode;
+  
 	return keyboard;
 };
 
+function getPrimaryAndSecondarySettingsForRegion(data, region) {
+  if (typeof data === 'string') {
+    if(typeof constants.colors[data] === 'undefined') throw 'invalid color for ' + region + ' region';
+    regionDef = {primary: {color: constants.colors[data], level: 0}, secondary: {color: constants.colors['off'], level: 0}};
+  } else if (typeof data === 'number') {
+    if(data < 0 || data >= constants.colors.length) throw 'invalid color for ' + region + ' region';
+    regionDef = {primary: {color: data, level: 0}, secondary: {color: constants.colors['off'], level: 0}};
+  } else if (typeof data === 'object') {
+    if (data.primary) {
+      regionDef = {primary: getColorAndIntensityLevel(data.primary, region, true)};
+    } else {
+      regionDef = {primary: getColorAndIntensityLevel(data, region, true)};
+    }
+    if (data.secondary) {
+      regionDef.secondary = getColorAndIntensityLevel(data.secondary, region);
+      //Default to off if both settings are unset
+      if (regionDef.secondary.color < 0 && regionDef.secondary.level < 0) regionDef.secondary = {color: constants.colors['off'], level: 0};
+      //Default to primary settings if either are unset
+      if (regionDef.secondary.color < 0) regionDef.secondary.color = regionDef.primary.color;
+      if (regionDef.secondary.level < 0) regionDef.secondary.level = regionDef.primary.level;
+    } else {
+      //Default secondary to off if unset
+      regionDef.secondary = {color: constants.colors['off'], level: 0};
+    }
+  } else 'invalid color for ' + region + ' region';
```

This does not return or respond with anything. This should throw an error as it will continue to line 80 and return an undefined variable.

`regionDef` should be defined at the top of the function as null.